File: src/reactive.js
// ============================================================
// reactive.js — Signal/memo/effect runtime
//
// Portable: no let/const, no arrows, no for-of, no Set,
// no .bind(), no template literals, no destructuring.
// Works in: Node, Bun, Deno, QuickJS, Duktape, Hermes,
// all browsers, V8/JSC/SpiderMonkey shell.
// ============================================================

var _currentObserver = null;
var _batchDepth = 0;
var _pendingEffects = [];

function createSignal(initialValue) {
  var value = initialValue;
  var subscribers = [];

  function read() {
    if (_currentObserver) {
      var found = false;
      for (var i = 0; i < subscribers.length; i++) {
        if (subscribers[i] === _currentObserver) { found = true; break; }
      }
      if (!found) subscribers.push(_currentObserver);
    }
    return value;
  }

  function write(nextValue) {
    if (typeof nextValue === "function") {
      value = nextValue(value);
    } else {
      value = nextValue;
    }
    var toRun = subscribers.slice();
    if (_batchDepth > 0) {
      for (var i = 0; i < toRun.length; i++) _pendingEffects.push(toRun[i]);
    } else {
      for (var i = 0; i < toRun.length; i++) toRun[i]._run();
    }
  }

  return [read, write];
}

function _Computation(fn) {
  this._fn = fn;
  this.value = undefined;
  this.dirty = true;
}

_Computation.prototype._run = function() {
  var prev = _currentObserver;
  _currentObserver = this;
  try {
    this.value = this._fn();
    this.dirty = false;
  } finally {
    _currentObserver = prev;
  }
  return this.value;
};

function createMemo(fn) {
  var comp = new _Computation(fn);
  comp._run();
  var origRun = comp._run;
  comp._run = function() {
    comp.dirty = true;
    return origRun.call(comp);
  };
  return function() {
    if (comp.dirty) comp._run();
    return comp.value;
  };
}

function createEffect(fn) {
  var comp = new _Computation(fn);
  comp._run();
}

function batch(fn) {
  _batchDepth++;
  try {
    fn();
  } finally {
    _batchDepth--;
    if (_batchDepth === 0) {
      var effects = _pendingEffects.slice();
      _pendingEffects = [];
      var seen = [];
      for (var i = 0; i < effects.length; i++) {
        var dup = false;
        for (var j = 0; j < seen.length; j++) {
          if (seen[j] === effects[i]) { dup = true; break; }
        }
        if (!dup) { seen.push(effects[i]); effects[i]._run(); }
      }
    }
  }
}

if (typeof exports !== "undefined") {
  exports.createSignal = createSignal;
  exports.createMemo = createMemo;
  exports.createEffect = createEffect;
  exports.batch = batch;
}
export { createSignal, createMemo, createEffect, batch };

File: src/reactive.py
# ============================================================
# reactive.py — Signal/memo/effect runtime
#
# Portable: CPython 3.6+, PyPy, MicroPython 1.19+.
# No dependencies. No set() (uses list for max compatibility).
# No f-strings, no walrus, no dataclasses.
# ============================================================

_observer = None
_batch_depth = 0
_pending = []


def signal(initial):
    state = [initial]
    subs = []

    def read():
        global _observer
        if _observer is not None:
            found = False
            for s in subs:
                if s is _observer:
                    found = True
                    break
            if not found:
                subs.append(_observer)
        return state[0]

    def write(val):
        global _batch_depth, _pending
        if callable(val):
            state[0] = val(state[0])
        else:
            state[0] = val
        to_run = list(subs)
        if _batch_depth > 0:
            _pending.extend(to_run)
        else:
            for s in to_run:
                s._run()

    return read, write


class _Computation:
    def __init__(self, fn):
        self.fn = fn
        self.value = None
        self.dirty = True

    def _run(self):
        global _observer
        prev = _observer
        _observer = self
        try:
            self.value = self.fn()
            self.dirty = False
        finally:
            _observer = prev
        return self.value


def memo(fn):
    comp = _Computation(fn)
    comp._run()
    orig_run = comp._run

    def trigger_run():
        comp.dirty = True
        return orig_run()
    comp._run = trigger_run

    def getter():
        if comp.dirty:
            comp._run()
        return comp.value
    return getter


def effect(fn):
    comp = _Computation(fn)
    comp._run()


def batch(fn):
    global _batch_depth, _pending
    _batch_depth += 1
    try:
        fn()
    finally:
        _batch_depth -= 1
        if _batch_depth == 0:
            effects = list(_pending)
            _pending = []
            seen = []
            for e in effects:
                dup = False
                for s in seen:
                    if s is e:
                        dup = True
                        break
                if not dup:
                    seen.append(e)
                    e._run()
